cpp/src/arrow/acero/source_node_test.cc

```diff
@@ -0,0 +1,109 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <gtest/gtest.h>
+
+#include "arrow/acero/exec_plan.h"
+#include "arrow/acero/map_node.h"
+#include "arrow/acero/options.h"
+#include "arrow/acero/test_nodes.h"
+
+namespace arrow {
+namespace acero {
+
+struct PauseThenStopNodeOptions : public ExecNodeOptions {
+  explicit PauseThenStopNodeOptions(int num_pass) : num_pass(num_pass) {}
+
+  int num_pass;
+};
+
+struct PauseThenStopNode : public MapNode {
+  static constexpr const char* kKindName = "PauseThenStopNode";
+  static constexpr const char* kFactoryName = "pause_then_stop";
+
+  static void Register() {
+    auto exec_reg = default_exec_factory_registry();
+    if (!exec_reg->GetFactory(kFactoryName).ok()) {
+      ASSERT_OK(exec_reg->AddFactory(kFactoryName, PauseThenStopNode::Make));
+    }
+  }
+
+  PauseThenStopNode(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                    std::shared_ptr<Schema> output_schema,
+                    const PauseThenStopNodeOptions& options)
+      : MapNode(plan, inputs, output_schema), num_pass(options.num_pass) {}
+
+  static Result<ExecNode*> Make(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                                const ExecNodeOptions& options) {
+    RETURN_NOT_OK(ValidateExecNodeInputs(plan, inputs, 1, kKindName));
+    auto bp_options = static_cast<const PauseThenStopNodeOptions&>(options);
+    if (bp_options.num_pass < 2) {
+      return Status::Invalid("num_pass must be at least 2");
+    }
+    return plan->EmplaceNode<PauseThenStopNode>(plan, inputs, inputs[0]->output_schema(),
+                                                bp_options);
+  }
+
+  const char* kind_name() const override { return kKindName; }
+  Result<ExecBatch> ProcessBatch(ExecBatch batch) override {
```

Comment:
Can you explain why this is supposed to work? Looks like the first batch will pause the input, then the second batch will stop? Could it happen that the second batch doesn't arrive because of the pause?

Reply:
This works because `PauseProducing` is not an immediate operation, which is useful for the current tester. In a debugging session, I observed that [the backpressure condition is detected immediately](https://github.com/apache/arrow/blob/a0d28deefd660adc6b14c4ac5fe4ed4175d4fedb/cpp/src/arrow/acero/source_node.cc#L213-L214) but [the loop](https://github.com/apache/arrow/blob/a0d28deefd660adc6b14c4ac5fe4ed4175d4fedb/cpp/src/arrow/acero/source_node.cc#L191) leads to [checking for termination](https://github.com/apache/arrow/blob/a0d28deefd660adc6b14c4ac5fe4ed4175d4fedb/cpp/src/arrow/util/future.h#L810) only after `StopProducing` is called. I suspect the non-immediacy is a consequence of managing using futures. @westonpace should be able to explain in more detail.

Reply:
That sounds about right.  In practice, at any given moment, we assume the source node is in the middle of obtaining the next batch (e.g. reading a batch from flight or reading a batch from the disk or some other operation).  This is the primary loop.  If we strip away the async stuff (and ignore synchronization) it would look something like...

```
void StartProducing() {
  while (true) {
    while (!paused_) {
      Batch next = source.Poll();
      if (!next) {
        // Finished reading the source
        return;
      }
      output_->InputReceived(next);
    }
    paused_cv_.wait();
  }
}

void PauseProducing() {
  paused_ = true;
}

void ResumeProducing() {
  paused_ = false;
  paused_cv_.notify_all();
}
```

Still, I would naively expect what Li said might technically be possible (e.g. 99% of the time when PauseProducing is called the above loop will be in `source.Poll` because that's the slow part. So you will get at least one batch after the pause.  However, if the timing is unlucky, maybe it is between `source.Poll` and the `!paused` check and you don't end up getting one more batch.

However, this is all running with `use_threads=false` which means the polling loop and the loop here are maybe all the same thread and so maybe the behavior is deterministic and correct.  I'd have to look further.

Reply:
@rtpsw Does this mean that the current test relies on one more batch to be delivered after PauseProducing is called, other wise this test will deadlock and never finishes? (i.e., we never call StopProducing)

Reply:
Although I suspect with the current implementation this is guaranteed, but I feel the fact that the test relies on that isn't great. It is possible to write the test in a more robust fashion? 

i.e., reusing the gated node idea, we can:
(1) Start a query with gated node closed so that back pressure is applied to source
(2) Call StopProducing 
(3) Validate that the execution is properly cancelled.

Reply:
I changed to an immediate stop after pause, which is a simpler way to avoid relying on this guarantee.